cli/lsp/language_server.rs

```diff
@@ -778,6 +819,210 @@ impl Inner {
+                &source,
+              ));
+            }
+          }
+          tsc::ScriptElementKind::ClassElement => {
+            if i.text != "<class>" {
+              code_lenses.push(i.to_code_lens(
+                &line_index,
+                &specifier,
+                &source,
+              ));
+            }
+          }
+          tsc::ScriptElementKind::InterfaceElement
+          | tsc::ScriptElementKind::TypeElement
+          | tsc::ScriptElementKind::EnumElement => {
+            code_lenses.push(i.to_code_lens(&line_index, &specifier, &source));
+          }
+          tsc::ScriptElementKind::LocalFunctionElement
+          | tsc::ScriptElementKind::MemberGetAccessorElement
+          | tsc::ScriptElementKind::MemberSetAccessorElement
+          | tsc::ScriptElementKind::ConstructorImplementationElement
+          | tsc::ScriptElementKind::MemberVariableElement => {
+            if let Some(parent) = &mp {
+              if parent.spans[0].start != i.spans[0].start {
+                match parent.kind {
+                  tsc::ScriptElementKind::ClassElement
+                  | tsc::ScriptElementKind::InterfaceElement
+                  | tsc::ScriptElementKind::TypeElement => {
+                    code_lenses.push(i.to_code_lens(
+                      &line_index,
+                      &specifier,
+                      &source,
+                    ));
+                  }
+                  _ => (),
+                }
+              }
+            }
+          }
+          _ => (),
+        }
+      }
+    });
+
+    self.performance.measure(mark);
+    Ok(Some(
+      Rc::try_unwrap(cl)
+        .map_err(|_| LspError::internal_error())?
+        .into_inner(),
+    ))
+  }
+
+  async fn code_lens_resolve(
+    &mut self,
+    params: CodeLens,
+  ) -> LspResult<CodeLens> {
+    let mark = self.performance.mark("code_lens_resolve");
+    if let Some(data) = params.data.clone() {
+      let code_lens_data: CodeLensData = serde_json::from_value(data)
+        .map_err(|err| LspError::invalid_params(err.to_string()))?;
+      let code_lens = match code_lens_data.source {
+        CodeLensSource::References => {
+          let line_index =
+            self.get_line_index_sync(&code_lens_data.specifier).unwrap();
+          let req = tsc::RequestMethod::GetReferences((
+            code_lens_data.specifier.clone(),
+            line_index.offset_tsc(params.range.start)?,
+          ));
+          let res =
+            self.ts_server.request(self.snapshot(), req).await.map_err(
+              |err| {
+                error!("Error processing TypeScript request: {}", err);
+                LspError::internal_error()
+              },
+            )?;
+          let maybe_references: Option<Vec<tsc::ReferenceEntry>> =
+            serde_json::from_value(res).map_err(|err| {
+              error!("Error deserializing response: {}", err);
+              LspError::internal_error()
+            })?;
```

Comment:
Not for this PR but in the future rather than calling `serde_json::from_value` after every call to `self.ts_server.request`, consider moving the `from_value` call into `request` by making it generic on the return type.

Reply:
Yeah, we do that a lot in quite a few places and yeah, it should be a generic.

Reply:
#9354 so we won't forget